chore(readme): contribute with local repro (#12074)
prisma/prisma: CONTRIBUTING.md

We also encourage you to join our sprawling [community](https://www.prisma.io/community) online, where you can discuss ideas, ask questions and get inspiration for what to build next.

> ## Table of Contents

Author: Not up-to-date and not needed anymore as github has a feature for this

pnpm run setup
```

> 💡 For Windows users: use the latest version of [Git Bash](https://gitforwindows.org/).

## Building packages when you make changes

In the root directory:

- `pnpm run setup` will install and build all the packages
- `pnpm -r run build` (-r for recursive) will build all the packages
- `pnpm -r run dev` (-r for recursive) will build all the packages, without running `tsc`

> - `pnpm run watch` will build any package that has been modified, without running `tsc` (Fastest)

Reviewer: Maybe

```suggestion
- `pnpm run watch` will continuously build any packages that have been modified, without running `tsc` (Fastest)
```

In a package directory, like `packages/client`:

- `pnpm run build` will build the package
- `pnpm run dev` will build the package without running `tsc` (Fastest)

> > 💡 Our builder is ESbuild

Reviewer: Nit: judging from its website and docs, the proper spelling seems to be esbuild in all lowercase, even as a proper noun in text.

> > 💡 Our builder is ESbuild

Author: ```suggestion
> 💡 Our builder is built on top of esbuild
```

pnpm run setup
```

> 💡 For Windows users: use the latest version of [Git Bash](https://gitforwindows.org/).

## Building packages when you make changes

In the root directory:

- `pnpm run setup` will install and build all the packages
- `pnpm -r run build` (-r for recursive) will build all the packages
- `pnpm -r run dev` (-r for recursive) will build all the packages, without running `tsc`
- `pnpm run watch` will continuously build any packages that have been modified, without running `tsc` (Fastest)

In a package directory, like `packages/client`:

- `pnpm run build` will build the package
- `pnpm run dev` will build the package without running `tsc` (Fastest)

> 💡 Our builder is built on top of esbuild

## Prisma Client

### First contribution

Create a reproduction folder for developing, trying a new feature, or a fix.

#### Setting up a locally-linked development folder

Set up a local project that will be linked to the local packages.

```sh
cd reproductions && pnpm install
# Copy a template from the reproduction folder
cp -r basic-sqlite my-repro && cd my-repro
# Ensure that the db and the schema are synced
pnpx prisma db push --skip-generate

> # Do some code changes, compile, then try it out

Reviewer 2: ```suggestion
# Do some code changes in the Prisma Client source code, re-generate the client, then try it out
```

> # Do some code changes, compile, then try it out
> pnpx prisma generate && pnpx ts-node index.ts

Reviewer 2: ```suggestion
# Re-generate the Prisma Client
pnpx prisma generate
# Run the script
pnpx ts-node index.ts
```
Maybe cleaner/simpler to cut into their own step

Author: I prefer to keep it in one command to be able to re-run the whole thing. Maybe I must make it more explicit that re-generate is needed every time?

Reviewer 2: Added a new suggestion to improve the comment

> # Do some code changes in the Prisma Client source code, re-generate the client, then try it out

Author: ```suggestion
# Do some code changes, always re-generate the client, then try it out
```

```

> 💡 This works best when compiling with `pnpm run watch` in the background.

> 💡 In any successful setup `pnpx prisma -v` should return version `0.0.0`.

<details>
  <summary><b>Alternatives</b></summary>
  
  #### Detailed steps for a locally-linked dev folder
  ```sh
  cd reproductions
  mkdir my-repro
  cd my-repro
  pnpm init -y
  pnpm add ../../prisma/packages/client

>   pnpm add -D ../../prisma/packages/cli

Reviewer 2: When executing this from the root folder of the git repo it will result in these lines in the package.json
```
  "dependencies": {
    "client": "link:../../prisma/packages/client"
  },
  "devDependencies": {
...
    "cli": "link:../../prisma/packages/cli",
```

And here it seems that `link:` is extra and causes `pnpx` to ask to install latest

>   pnpm add ../../prisma/packages/client
>   pnpm add -D ../../prisma/packages/cli

Reviewer 2: ```suggestion
  pnpm add ../../packages/client
  pnpm add -D ../../packages/cli
```

  pnpm add -D typescript ts-node
  pnpm add -D @types/node
  touch index.ts
  pnpx tsc --init
  pnpx prisma init
  # > Manually populate the schema.prisma
  # > Manually add 👇 to the generator block
  #   output = "../node_modules/.prisma/client"
  # > Manually populate the index.ts
  pnpx prisma db push --skip-generate

>   pnpx prisma generate && pnpx ts-node index.ts # Try it out

Reviewer 2: ```suggestion
  pnpx prisma generate
  pnpx ts-node index.ts
```

pnpm run setup
```

> 💡 For Windows users: use the latest version of [Git Bash](https://gitforwindows.org/).

## Building packages when you make changes

In the root directory:

- `pnpm run setup` will install and build all the packages
- `pnpm -r run build` (-r for recursive) will build all the packages
- `pnpm -r run dev` (-r for recursive) will build all the packages without running `tsc`
- `pnpm run watch` will build any package that has been modified without running `tsc` (Fastest)

In a package directory, like `packages/client`:

- `pnpm run build` will build the package
- `pnpm run dev` will build the package without running `tsc` (Fastest)

> 💡 Our builder is ESbuild

## Prisma Client

### First contribution

Create a reproduction folder for developing, trying a new feature, or a fix.

#### Setting up a locally-linked development folder

> We suggest that you create a `repros` folder outside of the `prisma` folder.  

Reviewer 2: This says `outside of the `prisma` folder.` but I see a repro folder is contained in this PR, meaning it's inside prisma repo. I'm confused now

Author: Yes, basically I don't want people to work with their repros inside of the main repo and this one is basically a template (for now only `basic-sqlite`) that they can copy outside and work in.

Author: Any ideas about how to clarify that?

Reviewer 2: Thinking about this:

It could also work to have a `reproductions` directory at the root like you added
- which we put in .gitignore but with one example and add a README
- we can also ignore changes in our GitHub Actions (so jobs are not triggered by a change in this directory) [like here](https://github.com/prisma/prisma/blob/main/.github/workflows/test.yml#L8) 
- we can point people (internally or externally or both) to create a repro by copying the example

Like this, there is no problem of having it inside the monorepo?

You can use this to link locally modified and developed `@prisma` packages.

1. Copy `cp -r repros ../repros && cd ../repros/basic-sqlite`
1. Initialize `pnpm install && pnpx prisma db push --skip-generate`
1. Develop `pnpx prisma generate && pnpx ts-node index.ts`

> 💡 This works best when running `pnpm run watch` in the background.

> 💡 In any successful setup `pnpx prisma -v` should show version `0.0.0`.

<details>
  <summary><b>Alternatives</b></summary>
  
  #### Detailed steps for a locally-linked dev folder

>   1. `cd ..`
>   1. `mkdir -p repros/my-repro`
>   1. `pnpm init -y`
>   1. `tsc --init`
>   1. `pnpm add ../../prisma/packages/client`
>   1. `pnpm add -D ../../prisma/packages/cli`
>   1. `pnpm add -D typescript ts-node`
>   1. `pnpm add -D @types/node`
>   1. `touch index.ts`
>   1. `pnpx prisma init`
>   1. Populate `schema.prisma`
>   1. `pnpx prisma db push --skip-generate`
>   1. Populate `index.ts`
>   1. `pnpx prisma generate && pnpx ts-node index.ts`

Reviewer 2: ```suggestion
```sh
cd ..
mkdir -p repros/my-repro
cd repros/my-repro
pnpm init -y
tsc --init
pnpm add ../../prisma/packages/client
pnpm add -D ../../prisma/packages/cli
pnpm add -D typescript ts-node
pnpm add -D @types/node
touch index.ts
pnpx prisma init
# Populate `schema.prisma`
pnpx prisma db push --skip-generate
# Populate `index.ts
pnpx prisma generate && pnpx ts-node index.ts
```
```

So someone can copy and paste the whole thing, instead of trying to copy one by one.
Also `cd repros/my-repro` was needed

Reviewer 2: The step `Populate `index.ts`` is not really clear, someone could thing it's a description of the step next to it.

Do you have an example of what to put into index.ts?

Reviewer 3: Added one backtick to make the code block suggestion render the code block inside correctly.

> #### Developing and working in the fixture folder
> 
> 1. `cd packages/client`
> 1. `ts-node fixtures/generate.ts ./fixtures/blog/ --skip-transpile`
> 1. `cd fixtures/blog`
> 1. `npx prisma db push --skip-generate` will create the database structure
> 1. `ts-node main`

Reviewer 3: ```suggestion
  #### Developing and working in the fixture folder
  
  1. `cd packages/client`
  1. `ts-node fixtures/generate.ts ./fixtures/blog/ --skip-transpile`
  1. `cd fixtures/blog`
  1. `npx prisma db push --skip-generate` will create the database structure
  1. `ts-node main`
```

>   1. `cd packages/client`
>   1. `ts-node fixtures/generate.ts ./fixtures/blog/ --skip-transpile`
>   1. `cd fixtures/blog`
>   1. `npx prisma db push --skip-generate` will create the database structure
>   1. `ts-node main`

Reviewer 2: ```suggestion
```sh
cd packages/client
ts-node fixtures/generate.ts ./fixtures/blog/ --skip-transpile
cd fixtures/blog
npx prisma db push --skip-generate # will create the database structure
ts-node main  
```
```

so it's possible to copy paste

pnpm run setup
```

> 💡 For Windows users: use the latest version of [Git Bash](https://gitforwindows.org/).

## Building packages when you make changes

In the root directory:

- `pnpm run setup` will install and build all the packages
- `pnpm -r run build` (-r for recursive) will build all the packages
- `pnpm -r run dev` (-r for recursive) will build all the packages without running `tsc`
- `pnpm run watch` will build any package that has been modified without running `tsc` (Fastest)

In a package directory, like `packages/client`:

- `pnpm run build` will build the package
- `pnpm run dev` will build the package without running `tsc` (Fastest)

> 💡 Our builder is ESbuild

## Prisma Client

### First contribution

Create a reproduction folder for developing, trying a new feature, or a fix.

#### Setting up a locally-linked development folder

We suggest that you create a `repros` folder outside of the `prisma` folder.  

> You can use this to link locally modified and developed `@prisma` packages.

Reviewer 2: ```suggestion
You can use this to link locally modified and developed Prisma packages.
```
CLI is not @ scoped so better to not have the npm scope here
